src/registries/menus/number_format_menu_registry.ts

```diff
@@ -1,66 +1,134 @@
+import { isDateTimeFormat, memoize } from "../../helpers";
 import { _t } from "../../translation";
-import { MenuItemRegistry } from "../menu_items_registry";
+import { Format } from "../../types";
+import { Registry } from "../registry";
 
-export const numberFormatMenuRegistry = new MenuItemRegistry();
+export const numberFormatMenuRegistry = new Registry<ACTION_FORMAT.NumberFormatActionSpec>();
 
 numberFormatMenuRegistry
   .add("format_number_automatic", {
     ...ACTION_FORMAT.formatNumberAutomatic,
+    id: "format_number_automatic",
     sequence: 10,
   })
   .add("format_number_plain_text", {
     ...ACTION_FORMAT.formatNumberPlainText,
+    id: "format_number_plain_text",
     sequence: 15,
     separator: true,
   })
   .add("format_number_number", {
     ...ACTION_FORMAT.formatNumberNumber,
+    id: "format_number_number",
     sequence: 20,
   })
   .add("format_number_percent", {
     ...ACTION_FORMAT.formatNumberPercent,
+    id: "format_number_percent",
     sequence: 30,
     separator: true,
   })
   .add("format_number_currency", {
     ...ACTION_FORMAT.formatNumberCurrency,
+    id: "format_number_currency",
     sequence: 40,
   })
   .add("format_number_currency_rounded", {
     ...ACTION_FORMAT.formatNumberCurrencyRounded,
+    id: "format_number_currency_rounded",
     sequence: 50,
   })
   .add("format_custom_currency", {
     ...ACTION_FORMAT.formatCustomCurrency,
+    id: "format_custom_currency",
     sequence: 60,
     separator: true,
   })
   .add("format_number_date", {
     ...ACTION_FORMAT.formatNumberDate,
+    id: "format_number_date",
     sequence: 70,
   })
   .add("format_number_time", {
     ...ACTION_FORMAT.formatNumberTime,
+    id: "format_number_time",
     sequence: 80,
   })
   .add("format_number_date_time", {
     ...ACTION_FORMAT.formatNumberDateTime,
+    id: "format_number_date_time",
     sequence: 90,
   })
   .add("format_number_duration", {
     ...ACTION_FORMAT.formatNumberDuration,
+    id: "format_number_duration",
     sequence: 100,
     separator: true,
   })
   .add("more_formats", {
     ...ACTION_FORMAT.moreFormats,
-    sequence: 110,
+    id: "more_formats",
+    sequence: 120,
   });
 
+export function getCustomNumberFormats(
+  env: SpreadsheetChildEnv
+): ACTION_FORMAT.NumberFormatActionSpec[] {
+  const defaultFormats = new Set(
+    numberFormatMenuRegistry
+      .getAll()
+      .map((f) => (typeof f.format === "function" ? f.format(env) : f.format))
+  );
+
+  const customFormats = new Map<Format, ACTION_FORMAT.NumberFormatActionSpec>();
+  for (const sheetId of env.model.getters.getSheetIds()) {
+    const cells = env.model.getters.getEvaluatedCells(sheetId);
+    for (const cellId in cells) {
+      const cell = cells[cellId];
+
+      if (cell.format && !customFormats.has(cell.format) && !defaultFormats.has(cell.format)) {
+        const formatType = getNumberFormatType(cell.format);
+        if (formatType !== "number") {
```

Comment:
the intention would be easier to understand if written the other way around
```suggestion
        if (formatType === "date" || formatType === "currency") {
```